### ISSUE TITLE
[Form] Wrong entity choices rendered with groups since 2.7

### ISSUE THREAD

**Issue by Soullivaneuh** [symfony/symfony#14915]
With this sample of code:

```php
$productsChoices = $this->getDoctrine()->getRepository('AppBundle:Product')->getChoicesGroupByCategory();
dump($productsChoices);

$form = $this->createFormBuilder()
    ->add('product', 'entity', [
        'class'     => 'AppBundle\Entity\Product',
        'choices'       => $productsChoices,
        'property'      => 'nameWithPrice',
    ])
    ->getForm()
;
```

And an extract of `dump($productsChoices);`:

```
array:23 [▼
  "Adresses IP" => array:1 [▼
    0 => Product {#3546 ▼
      -id: 12
      -name: "Adresse IP Failover"
      -ref: "xxx"
      -price: 42.0
      -isRecurrent: true
      -category: ProductCategory {#3641 ▶}
      -features: PersistentCollection {#3561 ▶}
    }
  ]
  "Bande-Passante" => array:3 [▼
    0 => Product {#3660 ▼
      -id: 65
      -name: "Bande passante"
      -ref: "xxx"
      -price: 42.0
      -isRecurrent: true
      -category: ProductCategory {#3663 ▶}
      -features: PersistentCollection {#3661 ▶}
    }
    1 => Product {#3668 ▶}
    2 => Product {#3670 ▶}
  ]
  "CDN" => array:3 [▼
    0 => Product {#3674 ▼
      -id: 24
      -name: "CDN Statique"
      -ref: "xxx"
      -price: 42.0
      -isRecurrent: true
      -category: ProductCategory {#3677 ▶}
      -features: PersistentCollection {#3675 ▶}
    }
    1 => Product {#3681 ▶}
    2 => Product {#3684 ▶}
  ]
```

Twig rendering:

```twig
{{ form(form) }}
```

I got this rendered select:

```html
<select id="form_product" name="form[product]" class=" form-control">
    <optgroup label="Adresses IP">
        <option value="18">VPN IpSec - 42.00 €</option>
    </optgroup>
    <optgroup label="Bande-Passante">
        <option value="18">VPN IpSec - 42.00 €</option>
        <option value="59">Ajout de disque - 42.00 €</option>
        <option value="58">Ajout de RAM - 42.00 €</option>
    </optgroup>
    <optgroup label="CDN">
        <option value="18">VPN IpSec - 42.00 €</option>
        <option value="59">Ajout de disque - 42.00 €</option>
        <option value="58">Ajout de RAM - 42.00 €</option>
    </optgroup>
    <!-- ... -->
</select>
```

For example, first value should not be "VPN IpSec" but "Adresse IP Failover".

This is a BC break and/or regressions because code was not modified and worked like a charm on `2.6`.

Using last dev-master Symfony version:

```bash
$ composer info -i | grep symfony/symfony
symfony/symfony                          2.7.x-dev e09874b   The Symfony PHP framework
```

Same error on `2.7.0` stable.

Regards

**Comment by Soullivaneuh** [comment]
Update: Using `group_by` option instead of custom choices values works:

```php
$form = $this->createFormBuilder()
    ->add('product', 'entity', [
        'class'     => 'Nexy\MainBundle\Entity\Product',
        'group_by'  => 'category.name',
        'property'  => 'nameWithPrice',
        'choice_translation_domain' => true,
    ])
    ->getForm()
;
```

But at least, I should not have to modify it because of BC rules.

**Comment by webmozart** [comment]
Hi @Soullivaneuh, thank you for reporting this issue! Could you please upload a fork of symfony-standard that reproduces your problem?

**Comment by Soullivaneuh** [comment]
To be quicker, I'll give you classes and code to reproduce it:

ProductCategory entity:
```php
/**
 * ProductCategory
 *
 * @ORM\Table(name="product_category")
 * @ORM\Entity
 */
class ProductCategory
{
    /**
     * @var int
     *
     * @ORM\Column(name="id", type="integer")
     * @ORM\Id
     * @ORM\GeneratedValue(strategy="AUTO")
     */
    private $id;

    /**
     * @var string
     *
     * @ORM\Column(name="name", type="string", length=64)
     */
    private $name;

    /**
     * @ORM\OneToMany(targetEntity="Product", mappedBy="category", cascade={"persist"})
     */
    private $products;

    public function __construct()
    {
        $this->products = new ArrayCollection();
    }

    public function getId()
    {
        return $this->id;
    }

    public function setName($name)
    {
        $this->name = $name;

        return $this;
    }

    public function getName()
    {
        return $this->name;
    }

    public function addProduct(Product $products)
    {
        $this->products[] = $products;

        return $this;
    }

    public function removeProduct(Product $products)
    {
        $this->products->removeElement($products);
    }

    public function getProducts()
    {
        return $this->products;
    }
}
```

ProductEntity:
```php
/**
 * Product
 *
 * @ORM\Table(name="product")
 * @ORM\Entity(repositoryClass="ProductRepository")
 */
class Product
{
    /**
     * @var int
     *
     * @ORM\Column(name="id", type="integer")
     * @ORM\Id
     * @ORM\GeneratedValue(strategy="AUTO")
     */
    private $id;

    /**
     * @var string
     *
     * @ORM\Column(name="name", type="string", length=64)
     */
    private $name;
    
    /**
     * @ORM\ManyToOne(targetEntity="ProductCategory", inversedBy="products")
     * @ORM\JoinColumn(name="category_id", referencedColumnName="id", nullable=false)
     */
    private $category;

    public function __construct()
    {
        $this->features = new ArrayCollection();
    }

    public function getId()
    {
        return $this->id;
    }

    public function setName($name)
    {
        $this->name = $name;

        return $this;
    }

    public function getName()
    {
        return $this->name;
    }

    public function setCategory(ProductCategory $category)
    {
        $this->category = $category;

        return $this;
    }

    public function getCategory()
    {
        return $this->category;
    }
}
```

ProductRepository:
```php
class ProductRepository extends EntityRepository
{
    public function getChoicesGroupByCategory()
    {
        $qb = $this->createQueryBuilder('p');
        $qb->select('p, c');

        $qb->leftJoin('p.category', 'c');
        $qb->orderBy('c.name', 'ASC')->addOrderBy('p.name', 'ASC');

        $products = $qb->getQuery()->execute();

        $choices = [];
        foreach ($products as $product) {
            $choices[$product->getCategory()->getName()][] = $product;
        }

        return $choices;
    }
}
```

And then, the code to reproduce this issue:

```php
$productsChoices = $this->getDoctrine()->getRepository('AppBundle:Product')->getChoicesGroupByCategory();

$form = $this->createFormBuilder()
    ->add('product', 'entity', [
        'class'     => 'AppBundle\Entity\Product',
        'choices'       => $productsChoices,
        'property'      => 'nameWithPrice',
    ])
    ->getForm()
;
```

Just tell me if I'm missing something.

Thanks.

**Comment by webmozart** [comment]
I know that this may seem useless to you, but please upload the code in a symfony-standard fork. This makes sure that I can actually reproduce the issue without losing too much time. After all, your bug may also stem from your configuration (which you don't post), the way you set up your application or similar.

**Comment by Soullivaneuh** [comment]
@webmozart Well, I'll take a look on my next free time.

**Comment by Soullivaneuh** [comment]
@webmozart Every go:

* Symfony fork with modification for bug reproducing: https://github.com/Soullivaneuh/symfony-standard/tree/issue-14915
* Compare from 2.7 branch: https://github.com/symfony/symfony-standard/compare/2.7...Soullivaneuh:issue-14915

Hope it's help.

**Comment by webmozart** [comment]
Thanks! I can confirm your problem. Investigating now.

**Comment by vinniecent** [comment]
I can confirm this bug in a Sliex application that uses the Symfony Form bundle.

The following array:
```
$choices = array(
	'Group 1' =>
		array(
			"id_1" => "sub option 1",
			"id_2" => "sub option 2"
		),
	'Group 2' =>
		array(
			"id_3" => "sub option 1",
			"id_4" => "sub option 2"
		)
);
```
...produces this html select element:
```
<select required="required" name="product_item[store_category]" id="product_item_store_category">
	<optgroup label="Group 1">
    	<option value="id_3">sub option 1</option>
    	<option value="id_4">sub option 2</option>
    </optgroup>
    <optgroup label="Group 2">
        <option value="id_3">sub option 1</option>
        <option value="id_4">sub option 2</option>
    </optgroup>
</select>
```

This bug is not in the v2.6.8 release.

**Comment by webmozart** [comment]
This was fixed in #15061.
